tools/oprt2/pkg/commandrunner/runner.go

```diff
@@ -0,0 +1,171 @@
+		r.logger = logger
+	}
+}
+
+func NewRunner(opts ...RunnerOption) *Runner {
+	r := &Runner{
+		logger: logging.DiscardLogger,
+	}
+
+	for _, opt := range opts {
+		opt(r)
+	}
+
+	return r
+}
+
+func (r *Runner) RegisterHook(h Hook) {
+	if h == nil {
+		return
+	}
+
+	r.hooks = append(r.hooks, h)
+}
+
+// Close runs all cleanup hooks. This should always run, even if the
+// command errors.
+func (r *Runner) Close(ctx context.Context) error {
+	errs := make([]error, 0, len(r.hooks))
+	for _, hook := range r.hooks {
+		r.logger.DebugContext(ctx, "running hook cleanup", "hook", hook.Name())
+		errs = append(errs, hook.Cleanup(ctx))
+	}
+
+	return errors.Join(errs...)
+}
+
+// Setup runs setup hooks. This will return after the first hook failure, or all hooks succeed.
+func (r *Runner) Setup(ctx context.Context) error {
+	for _, hook := range r.hooks {
+		r.logger.DebugContext(ctx, "running hook setup", "hook", hook.Name())
+		if err := hook.Setup(ctx); err != nil {
+			return fmt.Errorf("hook %q setup failed: %w", hook.Name(), err)
+		}
+	}
+
+	r.setupRan = true
+	return nil
+}
+
+// Run executes the provided command. If setup hooks have not been successfully called, they
+// will be ran prior to executing the command.
+// Cleanup hooks will not be called explicitly. The caller should `Close` the runner after
+// all associated command have been ran.
+func (r *Runner) Run(ctx context.Context, name string, args ...string) error {
+	// Ensure setup has ran successfully at least once
+	if !r.setupRan {
+		if err := r.Setup(ctx); err != nil {
+			return fmt.Errorf("setup failed: %w", err)
+		}
+	}
+
+	// Allow hooks to modify the command or arguments
+	for _, hook := range r.hooks {
+		r.logger.DebugContext(ctx, "running hook pre-command", "hook", hook.Name())
+		if err := hook.PreCommand(ctx, &name, &args); err != nil {
+			cmdString := buildCommandDebugString(exec.Command(name, args...))
+			return fmt.Errorf("hook %q pre-command failed for command %q: %w",
+				hook.Name(), cmdString, err)
+		}
+	}
+
+	// Build the command
+	cmd := exec.CommandContext(ctx, name, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+
+	// Allow hooks to modify the command (add env vars, args, etc.)
+	for _, hook := range r.hooks {
+		r.logger.DebugContext(ctx, "running command hook", "hook", hook.Name())
+		if err := hook.Command(ctx, cmd); err != nil {
+			return fmt.Errorf("hook %q command failed on command %q: %w", buildCommandDebugString(cmd),
+				hook.Name(), err)
+		}
+	}
+
+	cmdDebugString := buildCommandDebugString(cmd)
+	r.logger.DebugContext(ctx, "running command", "command", cmdDebugString)
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("command %q failed: %w", cmdDebugString, err)
+	}
+
+	return nil
+}
+
+// buildCommandDebugString builds a textual version of cmd for debug logging.
+// This should never be directly executed.
+func buildCommandDebugString(cmd *exec.Cmd) string {
+	if cmd == nil {
+		return ""
+	}
```

Comment:
Nit: It seems a little weird to be handling this nil case inside the function. Can we just ensure that the method doesn't get a nil command?